api/buckets/service.go

```diff
@@ -239,39 +466,57 @@ func (s *Service) SetPath(ctx context.Context, req *pb.SetPathRequest) (*pb.SetP
 		return nil, fmt.Errorf("get bucket: %s", err)
 	}
 	buckPath := path.New(buck.Path)
+	encKey := buck.GetEncKey()
 
 	remoteCid, err := cid.Decode(req.Cid)
 	if err != nil {
-		return nil, fmt.Errorf("remote cid is invalid: %s", err)
+		return nil, fmt.Errorf("invalid remote cid: %s", err)
 	}
 	remotePath := path.IpfsPath(remoteCid)
 
-	var dirpth path.Path
+	var dirpth path.Resolved
 	if req.Path == "" {
-		seed := make([]byte, 32)
-		if _, err := rand.Read(seed); err != nil {
+		sn, err := makeSeed(encKey)
+		if err != nil {
 			return nil, fmt.Errorf("generating new seed: %s", err)
 		}
-		dirpth, err = s.createBootstrappedPath(ctx, seed, remoteCid)
+		dirpth, err = s.createBootstrappedPath(ctx, sn, remoteCid, encKey)
 		if err != nil {
 			return nil, fmt.Errorf("generating bucket new root: %s", err)
 		}
-	} else {
-		dirpth, err = s.IPFSClient.Object().AddLink(ctx, buckPath, req.Path, remotePath, options.Object.Create(true))
-		if err != nil {
-			return nil, fmt.Errorf("adding folder: %s", err)
+		if encKey != nil {
+			if err = s.unpinNodeAndBranch(ctx, dirpth, encKey); err != nil {
+				return nil, fmt.Errorf("unpinning pinned root: %s", err)
+			}
+		} else {
+			if err = s.updateOrAddPin(ctx, buckPath, dirpth); err != nil {
+				return nil, fmt.Errorf("updating pinned root: %s", err)
+			}
 		}
-	}
-
-	if err = s.IPFSClient.Pin().Update(ctx, buckPath, dirpth); err != nil {
-		if err.Error() == pinNotRecursiveMsg {
-			if err = s.IPFSClient.Pin().Add(ctx, dirpth); err != nil {
-				return nil, fmt.Errorf("pinning new dag: %s", err)
+	} else {
+		if encKey != nil {
```

Comment:
This is probably the most complicated scenario...
1. Pull and encrypt a remote Unixfs node
2. Insert it into an already encrypted path